Separate Depthwidth and Lossguide growing policy in fast histogram

CodingCat opened the pull request:
it's a following PR of https://github.com/dmlc/xgboost/pull/4011 which separates depthwidth and loss guide growing policy in fast histogram per our discussion in https://github.com/dmlc/xgboost/issues/4077

closes https://github.com/dmlc/xgboost/issues/4077

@trivialfis @hcho3 @RAMitchell please help to review

hcho3 commented on jvm-packages/dev/build.sh:
```diff
@@ -17,5 +17,5 @@ rm /usr/bin/python
 
 # build xgboost
-cd /xgboost/jvm-packages;mvn package
+cd /xgboost/jvm-packages;ulimit -c unlimited;mvn package
```
Why `ulimit -c unlimited` here? Did you run into stack space issue?

CodingCat replied:
yes, it is a script used in docker container, I found (at least in my environment) that the stacktrace cannot be shown when debugging issues like code crashed in native layer

hcho3 commented:
@CodingCat Can you rebase against the latest master?

CodingCat commented:
@trivialfis @hcho3 @RAMitchell ping for review

CodingCat commented:
@trivialfis @hcho3 @RAMitchell ping for review

hcho3 commented:
Will review today

RAMitchell commented on src/common/column_matrix.h:
```diff
@@ -71,7 +71,6 @@ class ColumnMatrix {
                 double  sparse_threshold) {
     const auto nfeature = static_cast<bst_uint>(gmat.cut.row_ptr.size() - 1);
     const size_t nrow = gmat.row_ptr.size() - 1;
-
```
There seems to be a few unnecessary formatting changes in this PR. It makes it harder to review as we have to search through these carefully to see if anything is actually changed, as well as adding unnecessary history to these files.

RAMitchell commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
-  double time_build_hist = 0;
-  double time_evaluate_split = 0;
-  double time_apply_split = 0;
+  tstart = dmlc::GetTime();
```
I would like to see these timers replaced with the Monitor class.

CodingCat replied:
I have added the class

trivialfis commented:
Glanced the changes.  Will leave comments today.  Got into some troubles with my network access recently.  Sorry for the delay.

RAMitchell commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
   tstart = dmlc::GetTime();
-  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
-  time_init_data = dmlc::GetTime() - tstart;
+  for (int depth = 0; depth < param_.max_depth + 1; depth++) {
```
The logic inside this for loop contains many different cases and will be difficult to read and maintain. Can we further simplify this? If I understand correctly most of this code is evaluating whether to actually build the histogram or defer construction of the histogram and use the subtraction trick.

RAMitchell commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
+        }
+      }
+    }
+    // 1. sync histogram
```
If you have to number your comments to describe the flow of this function perhaps these numbered sections should be functions on their own.

RAMitchell commented on src/tree/updater_quantile_hist.h:
```diff
@@ -97,13 +97,16 @@ class QuantileHistMaker: public TreeUpdater {
         hist_builder_.BuildHist(gpair, row_indices, gmat, hist);
       }
-      this->histred_.Allreduce(hist.data(), hist_builder_.GetNumBins());
+      if (sync_hist) {
```
Is there some penalty for calling this unnecessarily?

hcho3 commented on src/tree/updater_quantile_hist.h:
```diff
@@ -218,13 +235,41 @@ class QuantileHistMaker: public TreeUpdater {
+       std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
+                           std::function<bool(ExpandEntry, ExpandEntry)>>;
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided;
```
I think we should suffix all class variables with `_` to differentiate them from local variables. Example: `qexpand_loss_guided_`.

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
+          // in distributed setting, we always calcuate from left child or root node
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          if (left_to_right_siblings.find(nid) != left_to_right_siblings.end()) {
```
Use `IsLeft()` here?

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      int nid = qexpand_depth_wise[k].nid;
+      this->InitNewNode(nid, gmat, gpair_h, *p_fmat, *p_tree);
+      // add constraints
+      if (right_to_left_siblings.find(nid) != right_to_left_siblings.end()) {
```
Use `IsRight()` here?

CodingCat commented:
@RAMitchell @hcho3  thank you very much for the review, I have addressed most of the comments (will continue working on the performance monitor class tmr )

trivialfis commented on jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala:
```diff
@@ -111,7 +111,31 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
   }
 
-  test("test with fast histo with monotone_constraints") {
+  test("test with fast histo with monotone_constraints (lossguide)") {
```
`fast hist` is renamed as `quantile hist`.

trivialfis commented on jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala:
```diff
@@ -111,7 +111,31 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
+    val training = buildDataFrame(Classification.train)
+    val testDM = new DMatrix(Classification.test.iterator)
+    val paramMap = Map("eta" -> "1",
+      "max_depth" -> "6", "silent" -> "1",
```
`silent` is deprecated, use `verbosity` instead.

trivialfis commented on src/common/hist_util.cc:
```diff
@@ -1,8 +1,6 @@
 /*!
- * Copyright 2017-2018 by Contributors
+ * Copyright 2019 by Contributors
```
2017~2019?

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+    const GHistIndexBlockMatrix &gmatb,
+    RegTree *p_tree,
+    const std::vector<GradientPair> &gpair_h) {
+  // 0. build histogram for the nodes in current level
```
If you want to outline the steps, please gather the descriptions into one place.  The next item is 500 lines away...

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      if (!node.IsRoot() && node.IsLeftChild() &&
+          (row_set_collection_[nid].Size() <
+           row_set_collection_[(*p_tree)[node.Parent()].RightChild()].Size())) {
+        hist_.AddHistRow(nid);
```
Please move
``` c++
        hist_.AddHistRow(nid);
        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
```
out of the testing branches, I think they are the same in every branch.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+        // in distributed setting, we always calcuate from left child or root node
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        if (!node.IsRoot() && node.IsLeftChild()) {
```
`node.IsLeftChild()` is true.  No need to test.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
   time_init_data = dmlc::GetTime() - tstart;
 
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1) <<
```
This check is duplicated many times.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
 
   tstart = dmlc::GetTime();
+
   this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
```
This is called in `Update`.  It might be duplicated call, please verify.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+            temp_qexpand_depth);
+    // 4. if qexpand is empty and temp_qexpand is empty, break
+    if (temp_qexpand_depth.empty()) {
+      qexpand_depth_wise_.clear();
```
Please move
``` c++
      qexpand_depth_wise_.clear();
      nodes_to_derive_.clear();
```
out of testing branches.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+    while (!qexpand_loss_guided_->empty()) {
```
Can we move this loop into `ExpandWithLossGuide`?

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+    ExpandWithDepthWidth(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+  }
 
   // set all the rest expanding nodes to leaf
```
Is this comment still valid?

hcho3 replied:
This comment should be moved to line 359

hcho3 replied:
NVM, let's just remove the comment

trivialfis commented on src/tree/updater_quantile_hist.h:
```diff
@@ -127,18 +141,18 @@ class QuantileHistMaker: public TreeUpdater {
-                    const HistCollection& hist,
-                    const DMatrix& fmat,
-                    RegTree* p_tree);
+                     const GHistIndexMatrix& gmat,
```
Please see these indents on Github, your editor might need some adjustments.

trivialfis commented on src/tree/updater_quantile_hist.h:
```diff
@@ -218,19 +255,33 @@ class QuantileHistMaker: public TreeUpdater {
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided_;
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node is which
```
> value is the node is which

Typo?

trivialfis commented on src/tree/updater_quantile_hist.h:
```diff
@@ -218,19 +255,33 @@ class QuantileHistMaker: public TreeUpdater {
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node is which
+    // provides the evidence for substracts
+    std::unordered_map<int, int> nodes_to_derive_;
```
I think there might be a better word than "derive".

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
-                                        RegTree* p_tree) {
-  double gstart = dmlc::GetTime();
+void QuantileHistMaker::Builder::SyncHistograms(
+    int &starting_index,
```
I should probably get clang-tidy working as soon as possible, there is no need to pass as reference.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
-      << "tree_method=hist does not support multiple roots at this moment";
+    << "tree_method=hist does not support multiple roots at this moment";
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```
So `num_roots` is partially supported?  @hcho3 Is there any plan to add full support?  Otherwise we might just remove it.

trivialfis commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+}
+
+void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+                                    const GHistIndexBlockMatrix& gmatb,
```
Is your editor mixing tab and space?

CodingCat commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      if (!node.IsRoot() && node.IsLeftChild() &&
+          (row_set_collection_[nid].Size() <
+           row_set_collection_[(*p_tree)[node.Parent()].RightChild()].Size())) {
+        hist_.AddHistRow(nid);
```
actually no, for example 

if this is a left child and its number of samples is larger than its sibling's, we should do nothing and defer it to the time when we visit the sibling

if we move these lines out of test branches, it will be added in anyway

CodingCat commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+    while (!qexpand_loss_guided_->empty()) {
```
I actually feel this loop is not necessary, @hcho3 ?

hcho3 replied:
@CodingCat This loop exists so that the queue `qexpand_loss_guided_` gets emptied properly at the end. Note that some choice of split evaluators may result in early termination of the algorithm, with a bunch of leaf nodes left in the queue. This loop ensures that the nodes actually get labelled as leaf.

CodingCat replied:
but for loss guide, we always jump out from the while loop https://github.com/dmlc/xgboost/blob/master/src/tree/updater_quantile_hist.cc#L140, do you mean in other future cases?

hcho3 replied:
You're actually right; the loop is currently not needed. See comments 
https://github.com/dmlc/xgboost/blob/9b917cda4ff61f0512cdba1c53bc8e55258a5d4a/src/tree/updater_quantile_hist.cc#L197-L199

hcho3 replied:
Let's just remove this loop, along with the 3-line comments?

trivialfis commented:
Seems the comments got a little bit messy.  It's because my machine got frozen during a review.  After rebooting, the previous review was not visible to me so I have to do it again.  After publishing it the previous somehow shows up ... If you similar comments, please ignore one of them.

CodingCat commented:
I think I have addressed all comments, ready for the next round of review

RAMitchell commented:
Is there some reason you didn't use the existing Monitor class?
https://github.com/dmlc/xgboost/blob/99a290489cb8e7109ff4f8f1161f098185097b25/src/common/timer.h#L41

CodingCat commented:
@RAMitchell  I was not aware of this class...but when I take a look at Monitor class there, I found it has several differences with the performance monitoring here, e.g. the format of the report and the log level, I am a bit hesitated about whether we should change the behavior of performance reporting either in hist or in other places using that class

I personally vote to unify the performance monitoring across updaters after this release

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
-      << "tree_method=hist does not support multiple roots at this moment";
+    << "tree_method=hist does not support multiple roots at this moment";
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```
Opened #4133 to deprecate `num_roots` parameter.

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+  CHECK_EQ(p_tree->param.num_roots, 1)
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
```
Fix the indentation here?

hcho3 replied:
```suggestion
    ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
```

hcho3 commented on src/tree/updater_quantile_hist.h:
```diff
@@ -114,6 +119,94 @@ class QuantileHistMaker: public TreeUpdater {
    protected:
+    /* tree growing policies */
+    struct ExpandEntry {
+        int nid;
```
Why 4 spaces indent here? We should stick to 2 spaces

hcho3 commented on src/tree/updater_quantile_hist.h:
```diff
@@ -114,6 +119,94 @@ class QuantileHistMaker: public TreeUpdater {
+    };
+
+    struct TreeGrowingPerfMonitor {
+        enum timer_name {INIT_DATA, INIT_NEW_NODE, BUILD_HIST, EVALUATE_SPLIT, APPLY_SPLIT};
```
2 spaces instead of 4

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  perf_monitor.UpdatePerfTimer(TreeGrowingPerfMonitor::timer_name::INIT_DATA);
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
```
Remove this, since we're deprecating `num_roots`

hcho3 commented on src/tree/updater_quantile_hist.h:
```diff
@@ -218,13 +334,20 @@ class QuantileHistMaker: public TreeUpdater {
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node which
+    // provides the evidence for substracts
+    std::unordered_map<int, int> nodes_for_substrack_trick;
```
Let's use correct spelling:
```suggestion
    std::unordered_map<int, int> nodes_for_subtraction_trick;
```

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        nodes_for_substrack_trick[(*p_tree)[node.Parent()].RightChild()] = nid;
+        sync_count++;
```
Did you mean to use `(*sync_count)++;` here?

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        nodes_for_substrack_trick[(*p_tree)[node.Parent()].LeftChild()] = nid;
+        sync_count++;
```
Did you mean to use `(*sync_count)++;` here?

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+        // root node
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        sync_count++;
```
Did you mean to use `(*sync_count)++;` here?

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+    RegTree *p_tree) {
+  perf_monitor.TickStart();
+  this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
+  // use substracttricks
```
```suggestion
  // use Subtraction Trick
```

hcho3 commented on src/tree/updater_quantile_hist.h:
```diff
@@ -218,13 +334,20 @@ class QuantileHistMaker: public TreeUpdater {
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided_;
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node which
```
```suggestion
    // key is the node id which should be calculated by Subtraction Trick, value is the node which
```

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      int left_id = (*p_tree)[nid].LeftChild();
+      int right_id = (*p_tree)[nid].RightChild();
+      temp_qexpand_depth->push_back(ExpandEntry(left_id,
+                                               p_tree->GetDepth(left_id), 0.0, (*timestamp)++));
```
```suggestion
                                                p_tree->GetDepth(left_id), 0.0, (*timestamp)++));
```

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
+      temp_qexpand_depth->push_back(ExpandEntry(left_id,
+                                               p_tree->GetDepth(left_id), 0.0, (*timestamp)++));
+      temp_qexpand_depth->push_back(ExpandEntry(right_id,
+                                               p_tree->GetDepth(right_id), 0.0, (*timestamp)++));
```
```suggestion
                                                p_tree->GetDepth(right_id), 0.0, (*timestamp)++));
```

hcho3 commented on src/tree/updater_quantile_hist.cc:
```diff
@@ -89,120 +88,275 @@ bool QuantileHistMaker::UpdatePredictionCache(
+  unsigned timestamp = 0;
+  int num_leaves = 0;
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```
Once this is merged, I will file a follow-up PR to deprecate `num_roots` parameter.

CodingCat commented:
thanks for the review @trivialfis @hcho3 @RAMitchell

Denisevi4 commented:
So depthwidth is now going to be per-level? :( 

That's unfortunate, because we use depthwidth and the per-nodeness of fast histogram to do feature selection - addding penalty to loss change for unused features. Once feature is used, no penalty is applied. With per-levelness it wouldn't work well especially for deep trees. Was it really worth it?

CodingCat commented:
@Denisevi4 it helps for reducing communication overhead in distributed training and serves as the base for improving multi-cores performance in the next release by increasing maximum parallelism

hcho3 commented:
@Denisevi4 The per-nodeness of fast histogram was not so good for performance, due to 1) extra syncing needed in distributed setting, and 2) lack of parallelism in multi-core CPU
